apisix/discovery/consul/init.lua

```diff
@@ -0,0 +1,418 @@
+        if not ok then
+            log.error("create ngx_timer_at got error: ", err)
+            return
+        end
+    end
+end
+
+
+local function format_consul_params(consul_conf)
+    local consul_server_list = core.table.new(0, #consul_conf.servers)
+    local args
+
+    if consul_conf.keepalive == false then
+        args = {}
+    elseif consul_conf.keepalive then
+        args = {
+            wait = consul_conf.timeout.wait, --blocked wait!=0; unblocked by wait=0
+            index = 0,
+        }
+    end
+
+    for _, v in pairs(consul_conf.servers) do
+        local scheme, host, port, path = unpack(http.parse_uri(nil, v))
+        if scheme ~= "http" then
+            return nil, "only support consul http schema address, eg: http://address:port"
+        elseif path ~= "/" or core.string.has_suffix(v, '/') then
+            return nil, "invalid consul server address, the valid format: http://address:port"
+        end
+
+        core.table.insert(consul_server_list, {
+            host = host,
+            port = port,
+            connect_timeout = consul_conf.timeout.connect,
+            read_timeout = consul_conf.timeout.read,
+            consul_sub_url = "/agent/services",
+            consul_watch_sub_url = "/catalog/services",
+            consul_server_url = v .. "/v1",
+            weight = consul_conf.weight,
+            keepalive = consul_conf.keepalive,
+            default_args = args,
+            index = 0,
+            fetch_interval = consul_conf.fetch_interval -- fetch interval to next connect consul
+        })
+    end
+
+    return consul_server_list
+end
+
+
+function _M.init_worker()
+    local consul_conf = local_conf.discovery.consul
+
+    if consul_conf.dump then
+        local dump = consul_conf.dump
+        dump_params = dump
+
+        if dump.load_on_init then
+            read_dump_services()
+        end
+    end
+
+    events = require("resty.worker.events")
+    events_list = events.event_list(
+            "discovery_consul_update_all_services",
+            "updating"
+    )
+
+    if 0 ~= ngx.worker.id() then
+        events.register(discovery_consul_callback, events_list._source, events_list.updating)
+        return
+    end
+
+    log.notice("consul_conf: ", json_delay_encode(consul_conf, true))
+    default_weight = consul_conf.weight
+    -- set default service, used when the server node cannot be found
+    if consul_conf.default_service then
+        default_service = consul_conf.default_service
+        default_service.weight = default_weight
+    end
+    if consul_conf.skip_services then
+        skip_service_map = core.table.new(0, #consul_conf.skip_services)
+        for _, v in ipairs(consul_conf.skip_services) do
+            skip_service_map[v] = true
+        end
+    end
+
+    local consul_servers_list, err = format_consul_params(consul_conf)
+    if err then
+        error(err)
+        return
```

Comment:
Can we remove the unreachable code here?

Reply:
the `format_consul_params` will return error if the config uri is not valid

Reply:
@Fabriceli 
Yes. I mean if the `error` is hit, the `return` here is not executed.

Reply:
done